src/ArgumentEscaper.cs

```diff
@@ -0,0 +1,207 @@
+namespace RunScript;
+
+/// <summary>
+/// Copy of the internal sdk class to escape command arguments updated to use <c>ReadOnlySpan</c>.
+/// </summary>
+/// <remarks>
+/// <seealso href="https://github.com/dotnet/sdk/blob/09b31215867d1ffe4955fd5b7cd91eb552d3632c/src/Cli/Microsoft.DotNet.Cli.Utils/ArgumentEscaper.cs"/>
+/// </remarks>
+internal static class ArgumentEscaper
+{
+    private const char Backslash = '\\';
+    private const char Caret = '^';
+    private const char NewLine = '\n';
+    private const char Quote = '"';
+    private const char Space = ' ';
+    private const char Tab = '\t';
+
+    /// <summary>
+    /// Undo the processing which took place to create string[] args in Main, so that the next process will receive the same string[] args.
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="args">List of arguments to escape.</param>
+    /// <returns>An escaped string of the supplied arguments.</returns>
+    public static string EscapeAndConcatenateArgArrayForProcessStart(IReadOnlyList<string>? args)
+    {
+        if (args is null)
+        {
+            return "";
+        }
+
+        return EscapeArgArray(args);
+    }
+
+    /// <summary>
+    /// Undo the processing which took place to create string[] args in Main, so that the next process will receive the same string[] args.
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="args">List of arguments to escape.</param>
+    /// <returns>A <c>cmd.exe</c> safe escaped string of the supplied arguments.</returns>
+    public static string EscapeAndConcatenateArgArrayForCmdProcessStart(IReadOnlyList<string>? args)
+    {
+        if (args is null)
+        {
+            return "";
+        }
+
+        return EscapeArgArrayForCmd(args);
+    }
+
+    /// <summary>
+    /// Undo the processing which took place to create string[] args in Main, so that the next process will receive the same string[] args.
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="arguments">List of arguments to escape.</param>
+    /// <returns>An escaped string of the supplied arguments.</returns>
+    private static string EscapeArgArray(IReadOnlyList<string> arguments)
+    {
+        var sb = new ValueStringBuilder(stackalloc char[256]);
+
+        for (var i = 0; i < arguments.Count; i++)
```

Comment:
I'm curious if there is any performance difference with enumerator. Also, then you could just make this IEnumerable. You are already reaching in via an interface so probably won't do much harm.
```csharp
var enumerator = arguments.GetEnumerator();
if (enumerator.MoveNext()) {
  var sb = new ValueStringBuilder(stackalloc char[256]);
  EscapeSingleArg(ref sb, enumerator.Current);

  while (enumerator.MoveNext()) {
    sb.Append(Space);
    EscapeSingleArg(ref sb, enumerator.Current);
  }
}
return string.Empty;
```

Reply:
Now that `args` is a `string[]` the `for` loop benchmarks the fastest here.